staging/src/k8s.io/client-go/openapi/openapitest/fake.go

```diff
@@ -0,0 +1,168 @@
+		return nil, errors.New(errStr)
+	}
+	values := map[string]openapi.GroupVersion{}
+	err = filepath.Walk(fullDirPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		filename := info.Name()
+		if !info.IsDir() && (filepath.Ext(filename) == ".json") {
+			apiFilename, err := apiFilepath(filename)
+			if err != nil {
+				// continue if filename is not required format
+				return nil
+			}
+			fullFilename := filepath.Join(fullDirPath, filename)
+			gvFile := FileOpenAPIGroupVersion{
+				Filepath: fullFilename,
+				Err:      nil,
+			}
+			values[apiFilename] = gvFile
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &FileOpenAPIClient{Values: values, Err: nil}, nil
+}
+
+// FileOpenAPIClient is a testing version implementing the
+// openapi.Client interface. This struct stores the hard-coded
+// Values returned by this file client, as well as a hard-coded
+// error if the user wants to force the return of an error.
+type FileOpenAPIClient struct {
+	Values map[string]openapi.GroupVersion
+	Err    error
+}
+
+// FileOpenAPIClient implements the openapi.Client interface.
+var _ openapi.Client = &FileOpenAPIClient{}
+
+// Paths returns the hard-coded map of the api server relative URL
+// path string to the GroupVersion swagger bytes. If the "err" field
+// is non-nil, this fake client returns the hard-coded error instead
+// of hard-coded "Values". An example Path string for apps/v1 GroupVersion
+// is:
+//
+//	apis/apps/v1
+func (f FileOpenAPIClient) Paths() (map[string]openapi.GroupVersion, error) {
+	if f.Err != nil {
+		return nil, f.Err
+	}
+	return f.Values, nil
+}
+
+// FileOpenAPIGroupVersion is a testing version implementing the
+// openapi.GroupVersion interface. This struct stores the full
+// filepath to the file storing the hard-coded GroupVersion bytes.
+// If the err is non-nil, then the error is returned by this
+// structs method.
+type FileOpenAPIGroupVersion struct {
+	Filepath string
+	Err      error
+}
+
+// FileOpenAPIGroupVersion implements the openapi.GroupVersion interface.
+var _ openapi.GroupVersion = &FileOpenAPIGroupVersion{}
+
+// Schemas returns the GroupVersion bytes at the stored filepath,
+// or an error if the "err" field is set. The passed content-type
+// is ignored, since we do not support protobuf.
+func (f FileOpenAPIGroupVersion) Schema(contentType string) ([]byte, error) {
+	if f.Err != nil {
+		return nil, f.Err
+	}
+	return ioutil.ReadFile(f.Filepath)
+}
+
+// apiFilepath is a helper function to parse a openapi filename
+// and transform it to the corresponding api relative url.
+// Examples:
+//
+//	apis__apps__v1_openapi.json -> apis/apps/v1
+//	apis__networking.k8s.io__v1alpha1_openapi.json -> apis/networking.k8s.io/v1alpha1
+//	api__v1_openapi.json -> api/v1
+func apiFilepath(filename string) (string, error) {
+	fields := strings.Split(filename, "__")
+	if len(fields) != 3 && len(fields) != 2 {
```

Comment:
`__` is a valid string to have in a group name so while this might work I am not a fan of doing it this way.

Looking inside these files I see that each spec only includes `paths` for its own groupversion. I dont think that'd be good to rely on either...

@apelisse WDYT for this. Stretch suggestions: should we update the openapi spec generator to include its source path in extension or field somewhere; or maybe change its naming scheme?

Reply:
Yeah, I agree with you, but on the other hand this function is literally doing the opposite of what the script does. The problem is that the conversion function used by the script is not bijective. I would keep this code like this until we actually fix the script (if we do).

EDIT: Though reading the code, this is obviously more complicated than the thing we do in `hack/update-openapi-spec.sh`